tests/Unit/Ads/AccountServiceTest.php

```diff
@@ -382,10 +386,7 @@ public function test_setup_account_step_link_merchant_no_merchant_id() {
 				]
 			);
 
-		$this->expectException( Exception::class );
-		$this->expectExceptionMessage( 'A Merchant Center account must be connected' );
-
-		$this->account->setup_account();
+		$this->assertEquals( [ 'id' => self::TEST_ACCOUNT_ID ], $this->account->setup_account() );
```

Comment:
Same as previous comment can we check for a status change, or that the functions for linking were never called.

Reply:
As clarified in https://github.com/woocommerce/google-listings-and-ads/pull/2205#discussion_r1489567034

It's not so much about the test passing or not. It's that the assertions in the test do not fully check the expected behaviour. For example:

1. We mock `get_ads_id` to always return a test ID
2. We mock `get_merchant_id` to always return a 0
3. We mock the state with the step `link_merchant` to be pending
4. We check that `setup_account` returns the matching test ID we mocked in step 1

However what we don't test in this scenario, is if it actually attempted to link or not. If we look at the actual function:
```php
	private function link_merchant_account() {
		if ( ! $this->options->get_ads_id() ) {
			throw new Exception( 'An Ads account must be connected' );
		}

		// Create link for Merchant and accept it in Ads.
		$this->container->get( Merchant::class )->link_ads_id( $this->options->get_ads_id() );
		$this->container->get( Ads::class )->accept_merchant_link( $this->options->get_merchant_id() );
	}
```
With the input we provided of a test Ads ID and a 0 Merchant ID, it would try to link them. The call to the function gets prevented higher up, but we aren't actually confirming in this test, that's what happened. So we need to confirm that those functions were never called, which we can do with something like:

```php
$this->merchant->expects( $this->never() )->method( 'link_ads_id' );
$this->ads->expects( $this->never() )->method( 'accept_merchant_link' );
```

Or alternatively we need to check for a status change, which might be harder to test.

Same goes for both the tests:
- test_setup_account_step_link_merchant_no_ads_id
- test_setup_account_step_link_merchant_no_merchant_id

If they aren't going to throw an exception like they previously did then they need to be testing something valid, not just passing a test.

Reply:
@mikkamp I agree with your point. I have made the changes to assert that the `link_ads_id ` and `accept_merchant_link ` methods are never called when return early.

Reply:
Thanks for taking a look @ankitrox 
I tried testing those changes and it seems there is still some room for improvement.

I'm wondering why the following code was removed:
```php
		$this->state->expects( $this->once() )
			->method( 'get' )
			->willReturn(
				[
					'link_ads' => [ 'status' => MerchantAccountState::STEP_PENDING ],
				]
			);
```

If we don't mock that state then it just returns an empty state array, so it won't perform any step. In a scenario like that it's obvious that it won't call any functions to link, but we do need it to actually handle the expected step, otherwise the test once again becomes invalid.
All the comments about these tests apply to both `Ads/AccountServiceTest.php` and `MerchantCenter/AccountServiceTest.php`, so they'll need to be refactored in both.

Also when running the tests I get the following two warnings:
```
Warning: "continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? in google-listings-and-ads/src/Ads/AccountService.php on line 175

Warning:  "continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? in google-listings-and-ads/src/MerchantCenter/AccountService.php on line 343
```

If that code is intended to continue the `foreach` loop then `continue 2` must be used (otherwise it's just breaking out of the `switch` statement and will still set the step to completed). Although to avoid confusion I'd suggest to add a comment there to explain what the expected behaviour is. Also would like to point out that setting the step to STEP_PENDING without calling `$this->state->update( $state )` does nothing as the state won't be stored for the next time it calls these steps (this is especially a problem if it's the last step it's handling).

Reply:
@ankitrox I looked at the updates you made to the `Ads\AccountServiceTest::test_setup_account_step_link_merchant_no_merchant_id` method and they looked pretty good to me. For consistency, I went ahead and made the same changes to the other account linking tests in both Account Service test suites in [961d2d9](https://github.com/woocommerce/google-listings-and-ads/pull/2205/commits/961d2d96e1ca8918488b08a3f8755b2c1644969f) and also made sure in each case that we are confirming that the method that links the accounts is never called if one of the account IDs is not present.